File: boardGeneration/src/papjuli/szokereso/Board.java
package papjuli.szokereso;

import java.util.ArrayList;
import java.util.List;

class Board {
    String[][] letters;
    int size;
    int timeSeconds;
    List<String> words;
    int score;

    void setScoreFromWords() {
        score = 0;
        for (String word: words) {
            int length = word.length();
            if (length == 2) {
                score += 1;
            } else {
                score += (length * length - 5 * length + 10) / 2;
            }
        }
    }

    String asJson() {
        StringBuilder result = new StringBuilder();
        result.append("{");
        result.append("\"size\": ");
        result.append(size);
        result.append(", \"timeSeconds\": ");
        result.append(timeSeconds);
        result.append(", \"letters\": [");

        boolean firstRow = true;
        for (String[] row : letters) {
            result.append(firstRow ? "[" : ",[");
            firstRow = false;
            boolean firstLetter = true;
            for (String letter : row) {
                result.append(firstLetter ? "\"" : ",\"");
                firstLetter = false;
                result.append(letter);
                result.append("\"");
            }
            result.append("]");
        }
        result.append("]");

        result.append(", \"totalScore\": ");
        result.append(score);

        result.append(", \"words\": [");
        boolean firstWord = true;
        for (String word : words) {
            if (!firstWord) result.append(",");
            firstWord = false;
            result.append("\"");
            result.append(word);
            result.append("\"");
        }

        result.append("]}");
        return result.toString();
    }

    static Board getExampleBoard() {
        Board board = new Board();
        board.size = 2;
        board.timeSeconds = 60;
        board.letters = new String[2][2];
        board.letters[0][0] = "É";
        board.letters[0][1] = "S";
        board.letters[1][0] = "K";
        board.letters[1][1] = "Ő";
        board.words = new ArrayList<>();
        board.words.add("ÉS");
        board.words.add("ÉK");
        board.words.add("KÉS");
        board.words.add("KÉSŐ");
        board.words.add("KŐ");
        board.words.add("ŐK");
        board.words.add("ŐS");
        return board;
    }
}

File: ts/board_generation/Board.ts
class Board {
  public letters: Array<Array<string>>;
  public words: Array<string>;
  public totalScore: number;
  public size: number;
  private timeSeconds: number;

  constructor (size: number, timeSeconds: number) {
    this.size = size;
    this.timeSeconds = timeSeconds;
    this.letters = Array(size);
    for (let i = 0; i < size; i++) {
      this.letters[i] = Array(size);
    }
    this.totalScore = 0;
  }
 
  public asJson(): string {
    return JSON.stringify(this);
  }

  public static fromJson(boardJson: string): Board {
    let boardObj = JSON.parse(boardJson);
    let board = new Board(boardObj.size, boardObj.timeSeconds);
    for (var prop in boardObj) board[prop] = boardObj[prop];
    return board;
  }

  public setWords(words: Set<string>): void {
    this.words = new Array<string>();
    for (let word of words) {
      this.words.push(word);
    }
    this.words.sort();
    this.setScoreFromWords();
  }
  
  public setScoreFromWords(): void {
    this.totalScore = 0;
    for (let word of this.words) {
      let length = word.length;
      if (length == 2) {
        this.totalScore += 1;
      } else {
        this.totalScore += Math.floor((length * length - 5 * length + 10) / 2);
      }
    }
  }

  public static getExampleBoard(): Board {
    let board = new Board(2, 60);
    board.letters[0][0] = "É";
    board.letters[0][1] = "S";
    board.letters[1][0] = "K";
    board.letters[1][1] = "Ő";
    board.words = new Array<string>();
    board.words.push("ÉS");
    board.words.push("ÉK");
    board.words.push("KÉS");
    board.words.push("KÉSŐ");
    board.words.push("KŐ");
    board.words.push("ŐK");
    board.words.push("ŐS");
    return board;
  }
}